feat: add le, eb, em, money, inv_free, PlayerInv model, and getOpenSlotsInContainer

joshieman06 opened the pull request:
many things added

DevScyu committed:
ci: update github actions for workflow [skip ci]

kristofbolyai committed:
fix: Fix multiple memory leaks (#652)

* fix: Greatly reduce memory allocated on render in ItemFavoriteFeature

* fix: Do a bit more calculation for much less memory usage

* fix: Remove old reskinned names and greatly reduce allocations in parsing of SpellType

* fix: Optimize emerald pouch parsing

Co-authored-by: Scyu_ <scyu@scyu.dev>

WynntilsBot committed:
chore(release): v0.0.1-alpha.8 [skip ci]

kristofbolyai committed:
fix: Fix multiple issues with mini-quest summary (#648)

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.9 [skip ci]

kristofbolyai committed:
feat: Add Guild Attack Timer overlay (and render attack timer on map) (#649)

* feat: Add Guild Attack Timer overlay

* feat: Render Guild Attack Timer on map

* fix: Make overlay smaller

* feat: Sort by time remaining

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.10 [skip ci]

kristofbolyai committed:
refactor: Abstract map screen code (#655)

WynntilsBot committed:
chore(release): v0.0.1-alpha.11 [skip ci]

kristofbolyai committed:
feat: Add Discoveries Screen (#646)

* feat: Add ability to download discoveries from api

* feat: Add Discoveries screen base

* feat: Add Discovery queries

* feat: Add all textures for discovery screen

* feat: Finish WynntilsDiscoveriesScreen

* feat: Show progress

* fix: Review changes

WynntilsBot committed:
chore(release): v0.0.1-alpha.12 [skip ci]

kristofbolyai committed:
feat: Auto message territory defense on attack (#650)

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.13 [skip ci]

kristofbolyai committed:
feat: Add AuraTimerOverlayFeature (#653)

* feat: Add AuraTimerOverlayFeature

* fix: Remove debug

WynntilsBot committed:
chore(release): v0.0.1-alpha.14 [skip ci]

kristofbolyai committed:
feat: Parse defense in GuildAttackTimerOverlayFeature (#656)

* feat: Parse defense in GuildAttackTimerOverlayFeature

* fix: Fix one way of faking

* fix: Move TimedSet

* fix: Move TimedSet to utils

WynntilsBot committed:
chore(release): v0.0.1-alpha.15 [skip ci]

kristofbolyai committed:
feat: Add GuildMapScreen (#657)

* feat: Add AdvancementUpdateEvent

* feat: Parse territories from advancements

* feat: Add GuildMapScreen

* fix: Use CustomColor

* fix: Add comment

WynntilsBot committed:
chore(release): v0.0.1-alpha.16 [skip ci]

kristofbolyai committed:
fix: Add missing texture files

WynntilsBot committed:
chore(release): v0.0.1-alpha.17 [skip ci]

P0keDev committed:
feat: Mob HP Formatting (#661)

* feat: Abbreviate mob health feature

* fix: feature registry formatting

* use number formatting tools we already have

Co-authored-by: P0ke <alextodaro@rocketmailcom>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.18 [skip ci]

P0keDev committed:
feat: item stat decimal setting (#663)

Co-authored-by: P0ke <alextodaro@rocketmailcom>

WynntilsBot committed:
chore(release): v0.0.1-alpha.19 [skip ci]

kristofbolyai committed:
fix: Fix discovery queries

WynntilsBot committed:
chore(release): v0.0.1-alpha.20 [skip ci]

kristofbolyai committed:
fix: Do not send data when player is in special states (#667)

WynntilsBot committed:
chore(release): v0.0.1-alpha.21 [skip ci]

P0keDev committed:
feat: player armor hiding (#664)

Co-authored-by: P0ke <alextodaro@rocketmailcom>

WynntilsBot committed:
chore(release): v0.0.1-alpha.22 [skip ci]

kristofbolyai committed:
fix: Render player shadows correctly when injecting nametags (#666)

WynntilsBot committed:
chore(release): v0.0.1-alpha.23 [skip ci]

Incompleteusern committed:
feat: Stencils and support for compass distance (#631)

* initial impl

* a bunch of fixes

* more changes

* fix?

* spotless

* inline

* fix: Fix quilt mixin

* fix: Remove alpha testing

* comments, invert logic

* add compass distance overlay

Co-authored-by: kristofbolyai <49001742+kristofbolyai@users.noreply.github.com>

DevScyu committed:
ci: update workflow

WynntilsBot committed:
chore(release): v0.0.1-alpha.24 [skip ci]

kristofbolyai committed:
chore: Rebuild for quilt

WynntilsBot committed:
chore(release): v0.0.1-alpha.25 [skip ci]

P0keDev committed:
fix: update spell cost ID ranges (#668)

Co-authored-by: P0ke <alextodaro@rocketmailcom>

WynntilsBot committed:
chore(release): v0.0.1-alpha.26 [skip ci]

kristofbolyai committed:
fix: Use separate buffers for depth and stencil (#670)

WynntilsBot committed:
chore(release): v0.0.1-alpha.27 [skip ci]

P0keDev committed:
fix: Mana bank bar bugs (#673)

WynntilsBot committed:
chore(release): v0.0.1-alpha.28 [skip ci]

P0keDev committed:
fix: account for both types of boss bar name packets (#674)

WynntilsBot committed:
chore(release): v0.0.1-alpha.29 [skip ci]

HalcyonSuoh committed:
render waypoint distance in the world (#669)

kristofbolyai committed:
fix: Fix stencil mask buffer not clearing

WynntilsBot committed:
chore(release): v0.0.1-alpha.30 [skip ci]

kristofbolyai committed:
fix: GuildTerritoryModel should account for partial updates (#675)

WynntilsBot committed:
chore(release): v0.0.1-alpha.31 [skip ci]

kristofbolyai committed:
fix: WynntilsDiscoveriesScreen should load web cache more often (#676)

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.32 [skip ci]

kristofbolyai committed:
fix: Do not expose MinecraftSchedulerManager queue (#677)

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.33 [skip ci]

kristofbolyai committed:
fix: Add missing render previews for some overlays (#678)

WynntilsBot committed:
chore(release): v0.0.1-alpha.34 [skip ci]

P0keDev committed:
fix: correct chat encoding/decoding of spell cost ids (#679)

Co-authored-by: P0ke <alextodaro@rocketmailcom>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.35 [skip ci]

kristofbolyai committed:
feat: Add player cursor to the guild map (#681)

WynntilsBot committed:
chore(release): v0.0.1-alpha.36 [skip ci]

kristofbolyai committed:
fix: Lootrun recording fixes (#680)

* fix: Fix recorded lootruns not rendering

* fix: Fix multiple issues with lootrun recording

WynntilsBot committed:
chore(release): v0.0.1-alpha.37 [skip ci]

kristofbolyai committed:
fix: Fix status overlay not parsing some statuses (#683)

WynntilsBot committed:
chore(release): v0.0.1-alpha.38 [skip ci]

kristofbolyai committed:
fix: Fix friend join redirect for archers

WynntilsBot committed:
chore(release): v0.0.1-alpha.39 [skip ci]

kristofbolyai committed:
feat: Add /server list up to list servers with uptimes (and show uptime with /server info) (#686)

* feat: Add uptime to /server info

* feat: Add /server list up to list servers with uptimes

WynntilsBot committed:
chore(release): v0.0.1-alpha.40 [skip ci]

P0keDev committed:
fix: Misc. spell cost ID fixes (#684)

WynntilsBot committed:
chore(release): v0.0.1-alpha.41 [skip ci]

kristofbolyai committed:
feat: Add Chat Tabs (#671)

* feat: Add base for ChatTabs (WIP)

* feat: More code for chat tabs

* fix: Post ScreenInitEvent even if super is not called

* feat: Add chat tabs

* fix: Adjust guild regex

* fix: AND filters

* fix: Fix many bugs

* fix: Fix local messages not working

* fix: More fixes

* fix: Make chatTabData final

* refactor: Change matching order

* fix: Fix some weird issue with event bus and ForkJoinPool

* fix: Rename to ClientsideMessageEvent

* fix: Add Clientside RecipientType and more fixes

* fix: Rename low priority to consuming (high priority)

* fix: Fix matching for consuming tabs

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>
Co-authored-by: Alex Todaro <3767283+P0keDev@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.42 [skip ci]

kristofbolyai committed:
fix: Add all ProfessionTypes (#688)

WynntilsBot committed:
chore(release): v0.0.1-alpha.43 [skip ci]

kristofbolyai committed:
fix: Fix a bug which chat tabs that made action bar messages disappear

WynntilsBot committed:
chore(release): v0.0.1-alpha.44 [skip ci]

kristofbolyai committed:
fix: Disable feature as soon as possible on crash

WynntilsBot committed:
chore(release): v0.0.1-alpha.45 [skip ci]

kristofbolyai committed:
fix: Auto-apply chat tabs on feature reload (#689)

WynntilsBot committed:
chore(release): v0.0.1-alpha.46 [skip ci]

kristofbolyai committed:
fix: fix chat tabs loading from config (#690)

* fix: Auto-apply chat tabs on feature reload

* fix: fix chat tabs loading from config

WynntilsBot committed:
chore(release): v0.0.1-alpha.47 [skip ci]

kristofbolyai committed:
fix: Fix chat items not parsing + feature crash when screenshot copy fails (#692)

WynntilsBot committed:
chore(release): v0.0.1-alpha.48 [skip ci]

kristofbolyai committed:
fix: Fix chat history resetting (#693)

WynntilsBot committed:
chore(release): v0.0.1-alpha.49 [skip ci]

kristofbolyai committed:
fix: Add correct dependency

kristofbolyai committed:
fix: Let the user know about failed item screenshots

WynntilsBot committed:
chore(release): v0.0.1-alpha.50 [skip ci]

kristofbolyai committed:
feat: Make chat tabs with unread messages yellow (#696)

WynntilsBot committed:
chore(release): v0.0.1-alpha.51 [skip ci]

Zatzou committed:
feat: Add chat timestamps (#697)

* add chat timestamps

* ci: spotless formatting [skip ci]

* Update common/src/main/resources/assets/wynntils/lang/en_us.json

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* Update common/src/main/resources/assets/wynntils/lang/en_us.json

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* fix issues

* Update common/src/main/resources/assets/wynntils/lang/en_us.json

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* fix ordering of the class

* Change the timestamping to lowest event priority

This fixes issues with the timestamping happening before things like redirection filtering

* fix: Make ChatMessageReceivedEvent setMessage not conflict with others features

* fix: Inline variable

Co-authored-by: Zatzou <Zatzou@users.noreply.github.com>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

kristofbolyai committed:
chore: Build

WynntilsBot committed:
chore(release): v0.0.1-alpha.52 [skip ci]

kristofbolyai committed:
fix: Do not render lootruns twice a frame

WynntilsBot committed:
chore(release): v0.0.1-alpha.53 [skip ci]

YouSoPugly committed:
Adds config for max distance that the waypoint's text will render (#691)

* Adds config for max distance that they waypoint's text will render

* ci: spotless formatting [skip ci]

* If 0 always render

* ci: spotless formatting [skip ci]

* fix: Fix i18n

Co-authored-by: YouSoPugly <YouSoPugly@users.noreply.github.com>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

kristofbolyai committed:
fix: Fix Figura crash and black config screen (#698)

* fix: Fix incompatible mixins causing crash on world render

* fix: Fix black figura config screen on world

* fix: Add credits

WynntilsBot committed:
chore(release): v0.0.1-alpha.54 [skip ci]

kristofbolyai committed:
Update README.md

kristofbolyai committed:
[ci skip] Fix formatting for readme

kristofbolyai committed:
feat: Add tier 10 emerald pouch to pouch guide (#700)

WynntilsBot committed:
chore(release): v0.0.1-alpha.55 [skip ci]

kristofbolyai committed:
fix: Fix colors disappearing when a message has chat items (#699)

WynntilsBot committed:
chore(release): v0.0.1-alpha.56 [skip ci]

san7890 committed:
Fixes LICENSE Link in ReadMe.md (#701)

Was pointing towards outdated LICENSE on non-existing branch.

kristofbolyai committed:
fix: Fix discovery screen api cache issues (#703)

WynntilsBot committed:
chore(release): v0.0.1-alpha.57 [skip ci]

kristofbolyai committed:
fix: Do not scan quest book on entering world (#706)

WynntilsBot committed:
chore(release): v0.0.1-alpha.58 [skip ci]

kristofbolyai committed:
feat: Add /bombbell command (#704)

* fix: Fix discovery screen api cache issues

* feat: Add /bombbell command

* fix: Fix coloring

* feat: Finish commands

WynntilsBot committed:
chore(release): v0.0.1-alpha.59 [skip ci]

kristofbolyai committed:
fix: Fix chat tabs resetting on config reload (#705)

WynntilsBot committed:
chore(release): v0.0.1-alpha.60 [skip ci]

Incompleteusern committed:
feat: Support overflowing health and mana in bars (#707)

* add overflowing health

* ci: spotless formatting [skip ci]

* remove unused variable

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

kristofbolyai committed:
fix: Fix chat items resetting text colors after they appear (#710)

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.61 [skip ci]

Incompleteusern committed:
fix: Catch all exceptions with readJson (#711)

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.62 [skip ci]

P0keDev committed:
fix: ignore duplicated bomb bell messages (#713)

Co-authored-by: P0ke <alextodaro@rocketmailcom>

WynntilsBot committed:
chore(release): v0.0.1-alpha.63 [skip ci]

Incompleteusern committed:
Unbind unregistered keybinds (#623)

* Unbind unregistered keybinds

* fix: Fix keybinds not loading in after feature reload

* finish?

Co-authored-by: kristofbolyai <49001742+kristofbolyai@users.noreply.github.com>

kristofbolyai committed:
fix: Rework how overlays positions work internally (and fix up some overlay positions) (#708)

* fix: Better alignment saving in OverlayManagerGUI, Right and Bottom aligned overlays now maintain their position better

* fix: Remove useless method in overlay

* fix: Fix up some weird overlay alignments

* fix: Move info box overlay further down

* fix: Revert unnecessary breaking change

* fix: Fix comment

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.64 [skip ci]

kristofbolyai committed:
feat: Overwrite Territory Map button to open Guild Map (#712)

WynntilsBot committed:
chore(release): v0.0.1-alpha.65 [skip ci]

kristofbolyai committed:
fix: Fix crash loops in feature exception (#709)

WynntilsBot committed:
chore(release): v0.0.1-alpha.66 [skip ci]

kristofbolyai committed:
fix: Make unused ability point messages dark aqua

WynntilsBot committed:
chore(release): v0.0.1-alpha.67 [skip ci]

kristofbolyai committed:
fix: Fix ability point coloring

WynntilsBot committed:
chore(release): v0.0.1-alpha.68 [skip ci]

joshieman06 committed:
feat: added mana_pct function (#718)

* created ManaPctFunction

* added ManaPctFunction to the registry

* added mana_pct name and description

WynntilsBot committed:
chore(release): v0.0.1-alpha.69 [skip ci]

joshieman06 committed:
feat: added bps and bps_xz (#721)

* add name and desc for bps and bps_xz

* Update CharacterFunctions.java

* registered bps and bps_xz

* ci: spotless formatting [skip ci]

* removed static import & converted to camel case

* ci: spotless formatting [skip ci]

* Update common/src/main/resources/assets/wynntils/lang/en_us.json

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* fixed function names

Co-authored-by: joshieman06 <joshieman06@users.noreply.github.com>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

bricefrisco committed:
fix: Check if update has already been downloaded before downloading again (#724)

* Check if update file already exists before downloading it again

* Added condition when UPDATE_PENDING is returned

* Moved 'sendSuccess' out of lambda expression

* Changed message sent on update command

* ci: spotless formatting [skip ci]

Co-authored-by: bricefrisco <bricefrisco@users.noreply.github.com>

joshieman06 committed:
fix: emerald pouch parsing when picking emeralds up (#723)

* fix: emerald pouch parsing when picking emeralds up

emerald pouches now register even when picking up emeralds

* Update WynnItemMatchers.java

* i forgot a semicolon

* ci: spotless formatting

* refactor: Return early after instanceof check

Co-authored-by: joshieman06 <joshieman06@users.noreply.github.com>
Co-authored-by: kristofbolyai <49001742+kristofbolyai@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.70 [skip ci]

DevScyu committed:
ci: update worflow [skip ci]

kristofbolyai committed:
chore: Rebuild for quilt

WynntilsBot committed:
chore(release): v0.0.1-alpha.71 [skip ci]

kristofbolyai committed:
feat: Redirect "not enough mana", healing and speed effect messages (#722)

* feat: Redirect "not enough mana" messages

* feat: Redirect healing messages

* feat: Redirect speed boost messages

* fix: Use .matches

Co-authored-by: Alex Todaro <3767283+P0keDev@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.72 [skip ci]

P0keDev committed:
fix: include savior's sacrifice in status overlay (#729)

Co-authored-by: P0ke <alextodaro@rocketmailcom>

WynntilsBot committed:
chore(release): v0.0.1-alpha.73 [skip ci]

P0keDev committed:
feat: prevent trades and duels in combat (#728)

Co-authored-by: P0ke <alextodaro@rocketmailcom>

WynntilsBot committed:
chore(release): v0.0.1-alpha.74 [skip ci]

kristofbolyai committed:
feat: Add command aliases feature (#714)

Co-authored-by: Alex Todaro <3767283+P0keDev@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.75 [skip ci]

kristofbolyai committed:
fix: Allow clicks on locked emerald pouch and ability tree slots (#726)

* fix: Allow clicks on locked emerald pouch and ability tree slots

* fix: Add a config for emerald pouch clicks

* fix: typo

Co-authored-by: Alex Todaro <3767283+P0keDev@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.76 [skip ci]

kristofbolyai committed:
fix: Label pois should not be behind ServicePois (#730)

Co-authored-by: Alex Todaro <3767283+P0keDev@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.77 [skip ci]

kristofbolyai committed:
chore: Update gradle [skip ci] (#727)

* chore: Update gradle

* chore: Update to gradle 7.6

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

kristofbolyai committed:
fix: RenderUtils should not have floating-point texture coordinates (#731)

WynntilsBot committed:
chore(release): v0.0.1-alpha.78 [skip ci]

kristofbolyai committed:
feat: Add a GUI to create custom map icons, and add helper buttons to maps (#720)

* feat: Add map icons

* feat: Add Custom Pois

* feat: Add map buttons

* fix: Add missing helper to main map

* feat: Add helper and resource color toggle to guild map

* fix: Fix mouseClicked handler

* feat: Add PoiCreationScreen

* fix: Use "the compass marker"

WynntilsBot committed:
chore(release): v0.0.1-alpha.79 [skip ci]

kristofbolyai committed:
fix: Fix AbilityTreeScrollFeature bug

WynntilsBot committed:
chore(release): v0.0.1-alpha.80 [skip ci]

kristofbolyai committed:
chore: Rebuild broken quilt version

WynntilsBot committed:
chore(release): v0.0.1-alpha.81 [skip ci]

kristofbolyai committed:
chore: Rebuild broken version #2

WynntilsBot committed:
chore(release): v0.0.1-alpha.82 [skip ci]

kristofbolyai committed:
chore: Rebuild broken version #3

WynntilsBot committed:
chore(release): v0.0.1-alpha.83 [skip ci]

kristofbolyai committed:
fix: Fix CI build tasks

WynntilsBot committed:
chore(release): v0.0.1-alpha.84 [skip ci]

kristofbolyai committed:
fix: Respect autoTrackQuestCoordinates config (#735)

WynntilsBot committed:
chore(release): v0.0.1-alpha.85 [skip ci]

kristofbolyai committed:
feat: Add gathering tool tier overlay (#732)

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.86 [skip ci]

kristofbolyai committed:
fix: Give keybinds their default value on registration (#740)

WynntilsBot committed:
chore(release): v0.0.1-alpha.87 [skip ci]

kristofbolyai committed:
fix: Try to preserve invalid config file on invalid load (#738)

WynntilsBot committed:
chore(release): v0.0.1-alpha.88 [skip ci]

kristofbolyai committed:
fix: Show custom pois on minimap (#737)

kristofbolyai committed:
fix: Save config on waypoint deletion (#739)

WynntilsBot committed:
chore(release): v0.0.1-alpha.89 [skip ci]

kristofbolyai committed:
feat: Add center button to main map (#734)

WynntilsBot committed:
chore(release): v0.0.1-alpha.90 [skip ci]

kristofbolyai committed:
fix: Show custom pois on minimap (#737)

WynntilsBot committed:
chore(release): v0.0.1-alpha.91 [skip ci]

kristofbolyai committed:
feat: Auto create waypoints for chests (#736)

WynntilsBot committed:
chore(release): v0.0.1-alpha.92 [skip ci]

joshieman06 committed:
added emerald functions and open inv slots

joshieman06 committed:
added em functions and open inv slots

joshieman06 committed:
Create PlayerInventoryModel.java

joshieman06 committed:
add getEmptySlotsInContainer

joshieman06 committed:
add le, eb, em and inv_free

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/PlayerInventoryModel.java:
```diff
@@ -0,0 +1,51 @@
+    public static void init() {}
+
+    @SubscribeEvent
+    public static void worldChange(WorldStateEvent e) {
```
I think its preferred to rename these methods to `onWorldChange`

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/PlayerInventoryModel.java:
```diff
@@ -0,0 +1,51 @@
+    public static void worldChange(WorldStateEvent e) {
+        InventoryMenu inventory = McUtils.inventoryMenu();
+        emeralds = ContainerUtils.getEmeraldCountInContainer(inventory);
+        openSlots = ContainerUtils.getEmptySlotsInContainer(inventory);
```
```java
        InventoryMenu inventory = McUtils.inventoryMenu();
        emeralds = ContainerUtils.getEmeraldCountInContainer(inventory);
        openSlots = ContainerUtils.getEmptySlotsInContainer(inventory);
```

is duplicated across all three methods, perhaps separate it into another method

joshieman06 replied:
how necessary is it? i cant work on this atm, will do later if needed.

kristofbolyai replied:
> how necessary is it? i cant work on this atm, will do later if needed. 

We want to reduce duplicate code every case we can

joshieman06 committed:
Update PlayerInventoryModel.java

kristofbolyai commented on common/src/main/java/com/wynntils/functions/CharacterFunctions.java:
```diff
@@ -112,6 +113,88 @@ public String getValue(String argument) {
+    public static class EmeraldsFunction extends DependantFunction<Integer> {
+        @Override
+        public Integer getValue(String argument) {
+            return PlayerInventoryModel.getCurrentEmeraldCount() % 4096 % 64;
```
No need to do both %, the latter is enough

joshieman06 replied:
ah because 4096 is multiple of 64

kristofbolyai commented on common/src/main/java/com/wynntils/functions/CharacterFunctions.java:
```diff
@@ -112,6 +113,88 @@ public String getValue(String argument) {
+        @Override
+        public Integer getValue(String argument) {
+            int ems = PlayerInventoryModel.getCurrentEmeraldCount();
+            return (ems - (ems % 4096)) / 4096;
```
Why not just divide? As this is int, it will work

kristofbolyai commented on common/src/main/java/com/wynntils/functions/CharacterFunctions.java:
```diff
@@ -112,6 +113,88 @@ public String getValue(String argument) {
+        @Override
+        public Integer getValue(String argument) {
+            int ems = PlayerInventoryModel.getCurrentEmeraldCount();
+            return (ems % 4096 - (ems % 4096 % 64)) / 64;
```
(ems - ems /4096) / 64

kristofbolyai commented on common/src/main/java/com/wynntils/wynn/model/PlayerInventoryModel.java:
```diff
@@ -0,0 +1,51 @@
+    public static void init() {}
+
+    @SubscribeEvent
+    public static void onWorldChange(WorldStateEvent e) {
```
On world state change, set your caches to some default value (0?).  No need to calculate anything here

kristofbolyai commented on common/src/main/java/com/wynntils/wynn/model/PlayerInventoryModel.java:
```diff
@@ -0,0 +1,51 @@
+        return emeralds;
+    }
+
+    public static int getOpenInvSlots() {
```
getFreeInventorySlots

kristofbolyai commented on common/src/main/java/com/wynntils/wynn/model/PlayerInventoryModel.java:
```diff
@@ -0,0 +1,51 @@
+public class PlayerInventoryModel extends Model {
+
+    private static int emeralds = -1;
+    private static int openSlots = -1;
```
freeSlots

kristofbolyai commented on common/src/main/java/com/wynntils/wynn/model/PlayerInventoryModel.java:
```diff
@@ -0,0 +1,51 @@
+
+public class PlayerInventoryModel extends Model {
+
+    private static int emeralds = -1;
```
I would rather have these as 0 by default.

kristofbolyai commented on common/src/main/java/com/wynntils/wynn/model/PlayerInventoryModel.java:
```diff
@@ -0,0 +1,51 @@
+    private static int emeralds = -1;
+    private static int openSlots = -1;
+
+    public static void init() {}
```
Set your caches to 0 on init too just to be safe

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -658,8 +662,12 @@
   "function.wynntils.horse_xp.description": "Your horse's current xp",
   "function.wynntils.horse_xp.name": "Horse XP",
+  "function.wynntils.inventoryFree.description": "Number of open slots in inventory",
+  "function.wynntils.inventoryFree.name": "Open Inventory Slots",
```
Free Inventory Slots

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -658,8 +662,12 @@
   "function.wynntils.horse_tier.name": "Horse Tier",
   "function.wynntils.horse_xp.description": "Your horse's current xp",
   "function.wynntils.horse_xp.name": "Horse XP",
+  "function.wynntils.inventoryFree.description": "Number of open slots in inventory",
```
free slots

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -658,8 +662,12 @@
+  "function.wynntils.inventoryFree.name": "Open Inventory Slots",
   "function.wynntils.level.description": "Your current combat level",
   "function.wynntils.level.name": "Lvl",
+  "function.wynntils.liquidEmerald.description": "Amount of liquid emeralds in inventory",
```
Amount of money in liquid emeralds in your inventory

Currently it sounds like this function counts LE items, not overall money

Also make sure to reword eb and em translations

kristofbolyai commented on common/src/main/java/com/wynntils/core/functions/FunctionManager.java:
```diff
@@ -317,6 +317,11 @@ public static void init() {
         registerFunction(new CharacterFunctions.XpPctFunction());
         registerFunction(new CharacterFunctions.BpsFunction());
         registerFunction(new CharacterFunctions.BpsXzFunction());
+        registerFunction(new CharacterFunctions.LiquidEmeraldFunction());
```
Put these in alphabetical order. Make sure to move bps too, they break the order atm

magicus replied:
How is this resolved? It still looks very much out of order.

kristofbolyai reviewed (commented):
This certainly looks better! Few little changes and it will be ready.

joshieman06 committed:
Update CharacterFunctions.java

joshieman06 committed:
Update en_us.json

kristofbolyai committed:
fix: Zoom further and faster (#733)

WynntilsBot committed:
chore(release): v0.0.1-alpha.93 [skip ci]

joshieman06 committed:
Merge branch 'alpha' into alpha

kristofbolyai committed:
feat: Parse coordinates from chat and inject clickable component and add /compass share (#742)

* fix: Fix bug in TerritoryCommand

* fix: ChatItemFeature should not do the same check twice

* feat: Parse coordinates from chat and inject clickable component

* feat: Add /compass share

WynntilsBot committed:
chore(release): v0.0.1-alpha.94 [skip ci]

san7890 committed:
feat: Redirect Blacksmith Sell/Scrap Messages to NotificationManager Overlay (#743)

* framework

* spotless

* regex update, add to feature registries

* bleaugh lets try it this way

* yellow means go

* another variation

* actually implements the counter

this java thing hurts my brain

* okay let's try an early return because of crashes

bleaugh

* spotless time

* wait maybe stripping formatting was the issue

* charat is 0 apparently

* last iterance of self-work

* bolyai fixes

* fingers crossed

* gracefully handle single item selling

would drive me mad if i didn't do this one

* better pluralizer

* final touches, remove debug

* Spotless Apply

* typo

* Update common/src/main/resources/assets/wynntils/lang/en_us.json

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* ci: spotless formatting

* simple code review

* Refactor to cut down on unused code

Keeps selling and scrapping in two separate "lanes" to reduce wasted resources.

Cuts down on stringbuilder excess.

* spotless

* review

removes chatformatting macros
patches up items glitch
patches up integer glitch

* even moar spotless

* uuuuuuuuuuuuuuuugh

* palantir over ideaj

* ci: spotless formatting

* remove debugging code

* code cleanup

* fix: Change matching to work with item rarities

* fix: This now fully works

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>
Co-authored-by: san7890 <san7890@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.95 [skip ci]

kristofbolyai committed:
feat: Remove Quilted Fabric API dependency (#744)

WynntilsBot committed:
chore(release): v0.0.1-alpha.96 [skip ci]

kristofbolyai committed:
fix: Fix auto chest not setting after there is an instance of a specific tier (#747)

WynntilsBot committed:
chore(release): v0.0.1-alpha.97 [skip ci]

kristofbolyai committed:
chore: Add Quiltflower decompiler support and greatly improve workspace/development documentation (#746)

* chore: Use LoomQuiltFlower

* chore: Update README.md to include details about workspace setup

* fix: Fix typo

Co-authored-by: Incompleteusern <58920010+Incompleteusern@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.98 [skip ci]

san7890 committed:
feat: Redirect Potion Stack Messages to NotificationManager Overlay (#750)

* initial commit

* cleans up string

* rename PouchRedirect to InventoryRedirect

* Add to resources

* quick spotless sanity

WynntilsBot committed:
chore(release): v0.0.1-alpha.99 [skip ci]

magicus committed:
Update to the latest version of Parchment MC mappings (#748)

san7890 committed:
feat: Redirect Horse Despawning Messages to NotificationManager Overlay  (#745)

WynntilsBot committed:
chore(release): v0.0.1-alpha.100 [skip ci]

san7890 committed:
feat: Adds Fast Travel and Seaskipper Waypoints to Map (#749)

WynntilsBot committed:
chore(release): v0.0.1-alpha.101 [skip ci]

kristofbolyai committed:
[ci skip] chore: Fix README.md Quiltflower command

bricefrisco committed:
feat: Render player POI's on minimap (#754)

* Created PlayerPoiBase and PlayerMainMapPoi

* Added PlayerMiniMapPoi and render on Minimap

* ci: spotless formatting

* Added parenthesis to avoid confusion

* Demystified magic numbers

* ci: spotless formatting

Co-authored-by: bricefrisco <bricefrisco@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.102 [skip ci]

kristofbolyai committed:
fix: Fix chat coordinate parsing

WynntilsBot committed:
chore(release): v0.0.1-alpha.103 [skip ci]

kristofbolyai committed:
fix: Support + in chat coordinate parsing

WynntilsBot committed:
chore(release): v0.0.1-alpha.104 [skip ci]

san7890 committed:
feat: Redirects Territory (Town) Entering/Leaving Messages to NotificationManager Overlay (#753)

WynntilsBot committed:
chore(release): v0.0.1-alpha.105 [skip ci]

kristofbolyai committed:
[ci skip] refactor: Maintain alphabetical order in FeatureRegistry

san7890 committed:
feat: Redirects Shaman "No Totems Nearby" message to OverlayManager (#758)

This was getting on my nerves, especially the fact that it lacks a period at the end.

WynntilsBot committed:
chore(release): v0.0.1-alpha.106 [skip ci]

san7890 committed:
feat: Redirects Potion Replaced/At-Limit Messages as notifications (#757)

WynntilsBot committed:
chore(release): v0.0.1-alpha.107 [skip ci]

kristofbolyai committed:
fix: Fix `allowClickOnEmeraldPouchInBlockingMode` config logic (#755)

Co-authored-by: Alex Todaro <3767283+P0keDev@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.108 [skip ci]

HalcyonSuoh committed:
feat: Add off-screen indicator to WorldWaypointDistanceFeature (#760)

* off screen indicator

* ci: spotless formatting

* Update common/src/main/java/com/wynntils/gui/screens/maps/MainMapScreen.java

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* improve readability, make setDynamicCompassLocation also take an icon

* ci: spotless formatting

Co-authored-by: HalcyonSuoh <HalcyonSuoh@users.noreply.github.com>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

WynntilsBot committed:
chore(release): v0.0.1-alpha.109 [skip ci]

kristofbolyai committed:
feat: Add I18n for UpdateCommand and UpdatesFeature (#762)

WynntilsBot committed:
chore(release): v0.0.1-alpha.110 [skip ci]

kristofbolyai committed:
fix: Make coordinate parsing stricter (#761)

WynntilsBot committed:
chore(release): v0.0.1-alpha.111 [skip ci]

magicus committed:
Merge branch 'main' into alpha

WynntilsBot committed:
chore(release): v0.0.1-alpha.112 [skip ci]

WynntilsBot committed:
Merged 'main' into 'alpha'.

WynntilsBot committed:
chore(release): v0.0.1-alpha.113 [skip ci]

magicus reviewed (changes requested):
We have changed the main development branch from `alpha` back to `main`. This means you unfortunately need to close this PR, and create a new one from the same branch on your fork, but targeting `main` instead. Sorry for the trouble. 😢

WynntilsBot committed:
Merged 'main' into 'alpha'.

WynntilsBot committed:
chore(release): v0.0.1-alpha.114 [skip ci]

kristofbolyai committed:
Merge branch 'alpha' into alpha

kristofbolyai committed:
fix: Fix merge

kristofbolyai committed:
fix: Review fixes

kristofbolyai committed:
feat: Use our new cache-model in InventoryEmeraldCountFeature

kristofbolyai committed:
fix: Optimize cache

magicus commented on common/src/main/java/com/wynntils/wynn/model/PlayerInventoryModel.java:
```diff
@@ -0,0 +1,73 @@
+
+    @SubscribeEvent
+    public static void onContainerSetEvent(ContainerSetContentEvent e) {
+        // Only update if the container is the player inventory
```
Will this event be sent out "post" updating the inventory? So that `updateCache()` which grabs the players current inventory will really see the new content? (Same question about onSlotSetEvent below)

kristofbolyai replied:
Fixed!

magicus committed:
Merge branch 'main' into alpha

kristofbolyai committed:
fix: Fix events to run post

kristofbolyai committed:
Merge remote-tracking branch 'joshieman06/alpha' into alpha

magicus commented:
Nice! Just the sorting left then...

kristofbolyai commented:
> Nice! Just the sorting left then...

The FunctionManager init does not have its registers sorted, so I left that out intentionally. We should just create a new PR to fix that.

magicus reviewed (approved):
:shipit:

joshieman06 commented:
epic thanks guys

magicus commented:
No thank you @joshieman06 for your contribution!
